server/src/test/java/org/elasticsearch/indices/recovery/RecoverySourceHandlerTests.java

```diff
@@ -905,6 +902,359 @@ void recoverFilesFromSourceAndSnapshot(ShardRecoveryPlan shardRecoveryPlan,
+                                                    BlobStoreIndexShardSnapshot.FileInfo snapshotFile,
+                                                    ActionListener<Void> listener) {
+                    assertThat(repository, is(equalTo(snapshotFilesToRecover.getRepository())));
+                    assertThat(indexId, is(equalTo(snapshotFilesToRecover.getIndexId())));
+                    assertThat(containsSnapshotFile(snapshotFilesToRecover, snapshotFile), is(equalTo(true)));
+                    String fileName = snapshotFile.metadata().name();
+
+                    if (randomBoolean()) {
+                        filesFailedToDownload.add(fileName);
+                        if (randomBoolean()) {
+                            listener.onFailure(randomFrom(new IOException("Failure"), new CorruptIndexException("idx", "")));
+                        } else {
+                            throw new RuntimeException("Unexpected exception");
+                        }
+                    } else {
+                        filesRecoveredFromSnapshot.add(fileName);
+                        listener.onResponse(null);
+                    }
+                }
+
+                @Override
+                public void writeFileChunk(StoreFileMetadata fileMetadata,
+                                           long position,
+                                           ReleasableBytesReference content,
+                                           boolean lastChunk,
+                                           int totalTranslogOps,
+                                           ActionListener<Void> listener) {
+                    filesRecoveredFromSource.add(fileMetadata.name());
+                    listener.onResponse(null);
+                }
+            };
+
+            RecoverySourceHandler handler = new RecoverySourceHandler(
+                shard,
+                recoveryTarget,
+                threadPool,
+                getStartRecoveryRequest(),
+                between(1, 16),
+                between(1, 4),
+                between(1, 4),
+                between(1, 4),
+                true,
+                recoveryPlannerService) {
+                @Override
+                void createRetentionLease(long startingSeqNo, ActionListener<RetentionLease> listener) {
+                    listener.onResponse(new RetentionLease("id", startingSeqNo, 0, "test"));
+                }
+            };
+
+            PlainActionFuture<RecoverySourceHandler.SendFileResult> future = PlainActionFuture.newFuture();
+            handler.recoverFilesFromSourceAndSnapshot(shardRecoveryPlan,
+                store,
+                mock(StopWatch.class),
+                future
+            );
+            future.actionGet();
+
+            filesRecoveredFromSource.removeAll(sourceFilesToRecover);
+            filesRecoveredFromSource.removeAll(filesFailedToDownload);
+            assertThat(filesRecoveredFromSource, is(empty()));
```

Comment:
I think this does not verify that we actually recover failed files from source, since this will pass as long as not files beyond the original and failed files are recovered from source.

I think `filesRecoveredFromSource.equals(sourceFilesToRecover + filesFailedToDownload)` is what we want to verify?